Issue 20846: Create facelet views completely in Java without the use of a facelet file.

nmittles opened the pull request:
for #20846

Create facelet views completely in Java without the use of a facelet file.
New FAT tests for first class view support in Faces 4.0.

nmittles committed:
Temporary save to switch branches.

nmittles committed:
Issue 20846: Create facelet views completely in Java without the use of a facelet file.

New FAT tests for first class view support in Faces 4.0.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/build.gradle:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
Shouldn't this be a range: `2022, 2023` ?

nmittles replied:
Fixed.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/ProgrammaticFaceletTests.java:
```diff
@@ -0,0 +1,104 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
```
This license is incorrect here, needs to be updated to the EPL 2 license:

```
/*******************************************************************************
 * Copyright (c) 2023 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 * 
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/
```

pnicolucci replied:
Same comment for any other file in this PR that has an outdated copyright header.

nmittles replied:
Updated this and others.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/ProgrammaticFaceletTests.java:
```diff
@@ -0,0 +1,104 @@
+ *
+ */
+@RunWith(FATRunner.class)
+public class ProgrammaticFaceletTests {
```
Have you given any thought to FULL vs LITE modes?

nmittles replied:
Moving all but basic to FULL.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_programmaticFaceletServer/server.xml:
```diff
@@ -0,0 +1,20 @@
+
+    <featureManager>
+        <feature>faces-4.0</feature>
+        <feature>cdi-4.0</feature>
```
no need for the `cdi-4.0` feature as `faces-4.0` automatically enables `cdi-4.0` currently.

nmittles replied:
Updated.

pnicolucci replied:
Doesn't look like it got updated, I still see cdi-4.0.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/welcome.xhtml:
```diff
@@ -0,0 +1,16 @@
+            <h:commandButton id="button" value="Do action" />
+        </h:form>
+    </h:body>
+</html>
```
new line

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/WEB-INF/beans.xml:
```diff
@@ -0,0 +1,16 @@
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
+       bean-discovery-mode="all">
+</beans>
```
Do we even need a beans.xml here? If so we should use the latest version rather than 1.1.

nmittles replied:
Yes, otherwise I see CDI/Weld errors such as:
` WELD-001408: Unsatisfied dependencies`

I am updating the version.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/.classpath:
```diff
@@ -10,6 +10,8 @@
 	<classpathentry kind="src" path="test-applications/AjaxRenderExecuteThisTest.war/src"/>
 	<classpathentry kind="src" path="test-applications/AcceptInputFileTest.war/src"/>
+	<classpathentry kind="src" path="test-applications/ProgrammaticFaceletTests.war/src"/>
+	<classpathentry kind="src" path="test-applications/WebSocket.war/src"/>
```
Is the WebSocket.war/src actually part of this PR? I didn't see it.

nmittles replied:
This will be part of the websocket PR.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd:
```diff
@@ -23,7 +23,9 @@ src: \
-    test-applications/MultipleInputFileTest.war/src
+    test-applications/MultipleInputFileTest.war/src,\
+    test-applications/ProgrammaticFaceletTests.war/src,\
+    test-applications/WebSocket.war/src
```
This isn't part of this PR correct? Can this be cleaned up?

nmittles replied:
Correct. This will be part of the websocket PR.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd:
```diff
@@ -39,4 +41,8 @@ javac.target: 11
-	io.openliberty.jakarta.annotation.2.1;version=latest
+	io.openliberty.jakarta.annotation.2.1;version=latest,\
+	xml-apis:xml-apis;version='1.4.01',\
+	com.ibm.websphere.javaee.el.3.0,\
```
We should be using the latest Jakarta Expression Language 5.0 here.

pnicolucci replied:
Also the xml-apis has been removed from this bnd a couple of times so lets not add it back in :)

nmittles replied:
Removed 'com.ibm.websphere.javaee.el.3.0,\' as it was unnecessary. jakarta EL 5.0 is added below.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/build.gradle:
```diff
@@ -24,17 +24,25 @@
     requiredLibs 'xalan:xalan:2.7.2'
+    requiredLibs 'xml-apis:xml-apis:1.4.01'
     requiredLibs project(':com.ibm.ws.org.apache.commons.io')
+    requiredLibs project(':com.ibm.ws.org.slf4j.api')
```
are any of these updates necessary for this PR?

nmittles replied:
They will be used in a future PR for websocket testing.  Leaving them in after discussion.

volosied commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/src/io/openliberty/org/apache/faces40/fat/programmaticfacelettests/BasicViewFacelet.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ *
+ */
+@View("/BasicView.xhtml")
```
This works, but can you escape some of the text?  The rendered web page displays the html tags: 
```
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:h="http://xmlns.jcp.org/jsf/html">
Hello World!
</html>
```

`output.getAttributes().put("escape", false);`

volosied replied:
@pnicolucci  or should we leave it as is?   We might need to go to the community since myfaces doesn't escape but mojarra does.

volosied replied:
Update -- I was slightly incorrect above. MyFaces escapes special characters, so the html elements are visible on the page. Mojarra doesn't escape, so they don't face this issue. This should likely be brought up as a spec issue.

volosied replied:
Talked to Paul -- leave things as is. We can follow up on this later as I think this escape issue needs to resolved by the spec.

volosied commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/ProgrammaticFaceletTests.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    @Test
+    public void testBasicView() throws Exception {
+        HttpUtils.findStringInReadyUrl(server, "/" + PROGRAMMATIC_FACELET_TEST_APP_NAME + "/BasicView.xhtml", "Hello World!");
```
I don't know too much about HttpUtils. If the text is not found, is the page logged?   Could logging be added?

nmittles replied:
Yes, the page is logged as follows:
`The response did not contain "[Hello World!ABC]". Full output is:" &lt;html xmlns=&quot;http://www.w3.org/1999/xhtml&quot; xmlns:h=&quot;http://xmlns.jcp.org/jsf/html&quot;&gt;<head></head><body><form id="testForm" name="testForm" method="post" action="/ProgrammaticFaceletTests/BasicView.xhtml" enctype="application/x-www-form-urlencoded">Hello World!<input type="hidden" name="testForm_SUBMIT" value="1" /><input type="hidden" name="jakarta.faces.ViewState" id="j_id__v_0:jakarta.faces.ViewState:1" value="ZTAwN2U2YWIzZTU1ZWEwZDAwMDAwMDAx" autocomplete="off" /></form></body>&lt;/html&gt; ". expected:<0> but was:<1>
`

nmittles committed:
Code review updates 1.

nmittles commented:
#build

nmittles commented:
#libby

volosied commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/WEB-INF/beans.xml:
```diff
@@ -9,8 +9,8 @@
  * Contributors:
  *     IBM Corporation - initial API and implementation
```
Should this also be v2.0?

nmittles replied:
Yes, fixed it under build instead of publish, along with missing server others.  They should be fixed now.

volosied commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd:
```diff
@@ -39,4 +40,6 @@ javac.target: 11
-	io.openliberty.jakarta.annotation.2.1;version=latest
+	io.openliberty.jakarta.annotation.2.1;version=latest,\
+	io.openliberty.jakarta.expressionLanguage.5.0
+	
```
New Line

nmittles replied:
Fixed.

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/HtmlAndJava.xhtml:
```diff
@@ -0,0 +1,22 @@
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+
```
Wrong copyright header here as well

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/WEB-INF/web.xml:
```diff
@@ -0,0 +1,34 @@
+    Copyright (c) 2023 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
```
Wrong copyright header

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/test-applications/ProgrammaticFaceletTests.war/resources/welcome.xhtml:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
```
We should add a copyright header

nmittles replied:
added

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_programmaticFaceletServer/bootstrap.properties:
```diff
@@ -0,0 +1,17 @@
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
```
wrong copyright header

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_programmaticFaceletServer/osgi.debug.options:
```diff
@@ -0,0 +1,12 @@
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
```
wrong copyright header

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/publish/servers/faces40_programmaticFaceletServer/server.xml:
```diff
@@ -0,0 +1,20 @@
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
```
wrong copyright header

nmittles committed:
Code review comments 2.

Accidentally fixed copyright headers under build dir instead of publish.  Fixed missing new line.

nmittles commented:
#libby

nmittles commented:
#build

nmittles commented:
#build

nmittles commented:
#build

nmittles commented:
#build

nmittles commented:
#libby

pnicolucci commented on dev/io.openliberty.org.apache.myfaces.4.0_fat/.classpath:
```diff
@@ -11,6 +11,8 @@
 	<classpathentry kind="src" path="test-applications/AcceptInputFileTest.war/src"/>
 	<classpathentry kind="src" path="test-applications/AnnotationLiteralsTest.war/src"/>
+        <classpathentry kind="src" path="test-applications/ProgrammaticFaceletTests.war/src"/>
+	<classpathentry kind="src" path="test-applications/ProgrammaticFaceletTests.war/src"/>
```
This doesn't need to be declared twice.

nmittles committed:
Fix duplicate merged line.

nmittles commented:
#libby
